File: 4day/bank_system/managers/account_manager.py
"""
계좌 관리 매니저 클래스
Java의 AccountManager 클래스를 Python으로 변환
"""

from datetime import datetime
from typing import Optional, List
from ..database import get_database_connection, SQLQueries
from ..entities.account import Account
from ..entities.transaction import Transaction
from ..helpers.input_helper import InputHelper
from ..helpers.validation_helper import ValidationHelper
from ..utils.bank_utils import BankUtils
from ..utils.interest_calculator import InterestCalculator


class AccountManager:
    """계좌 관련 기능을 관리하는 클래스"""
    
    def __init__(self, user_manager=None, transaction_manager=None):
        """AccountManager 초기화"""
        self.db = get_database_connection()
        self.validator = ValidationHelper()
        self.input_helper = InputHelper()
        self.user_manager = user_manager
        self.transaction_manager = transaction_manager
        
        # InputHelper에 AccountManager 설정
        self.input_helper.set_account_manager(self)
    
    def set_transaction_manager(self, transaction_manager):
        """TransactionManager 설정"""
        self.transaction_manager = transaction_manager
    
    def account_exists(self, account_id: str) -> bool:
        """
        계좌 존재 여부 확인
        
        Args:
            account_id: 확인할 계좌번호
            
        Returns:
            bool: 계좌 존재 여부
        """
        try:
            results = self.db.execute_query(
                "SELECT COUNT(*) as count FROM accounts WHERE account_id = :account_id",
                {'account_id': account_id}
            )
            
            if results and len(results) > 0:
                count = results[0].get('COUNT', results[0].get('count', 0))
                return count > 0
            return False
            
        except Exception as e:
            print(f"계좌 존재 확인 오류: {e}")
            return False
    
    def is_my_account(self, account_id: str, user_id: str) -> bool:
        """
        본인 계좌 여부 확인
        
        Args:
            account_id: 확인할 계좌번호
            user_id: 사용자 ID
            
        Returns:
            bool: 본인 계좌 여부
        """
        try:
            results = self.db.execute_query(
                "SELECT COUNT(*) as count FROM accounts WHERE account_id = :account_id AND user_id = :user_id",
                {'account_id': account_id, 'user_id': user_id}
            )
            
            if results and len(results) > 0:
                count = results[0].get('COUNT', results[0].get('count', 0))
                return count > 0
            return False
            
        except Exception as e:
            print(f"계좌 소유권 확인 오류: {e}")
            return False
    
    def create_account(self, user_id: str) -> bool:
        """
        계좌 생성
        
        Args:
            user_id: 계좌 소유자 ID
            
        Returns:
            bool: 계좌 생성 성공 여부
        """
        try:
            print("\n[계좌 생성]")
            print("=" * 50)
            
            # 계좌 정보 입력
            account_name = self.input_helper.input("계좌명: ")
            account_type = self.input_helper.input_menu_choice(
                "계좌 종류 (1.보통예금 2.정기예금 3.적금): ",
                ['1', '2', '3']
            )
            
            # 계좌 종류 매핑
            type_map = {'1': '보통예금', '2': '정기예금', '3': '적금'}
            account_type = type_map[account_type]
            
            account_password = self.input_helper.input_account_password()
            initial_balance = self.input_helper.input_amount("초기 입금액: ")
            
            # 계좌번호 생성
            account_id = BankUtils.generate_account_number()
            if not account_id:
                print("계좌번호 생성에 실패했습니다.")
                return False
            
            # 이자율 설정
            interest_rate = InterestCalculator.get_interest_rate_by_type(account_type)
            
            # 계좌 객체 생성
            account = Account.create_account_with_interest(
                account_id=account_id,
                account_name=account_name,
                account_type=account_type,
                account_password=account_password,
                balance=initial_balance,
                user_id=user_id,
                interest_rate=interest_rate
            )
            
            # DB에 저장
            if self.save_account(account):
                print(f"✅ 계좌가 생성되었습니다!")
                print(f"계좌번호: {account_id}")
                print(f"계좌명: {account_name}")
                print(f"계좌종류: {account_type}")
                print(f"이자율: {InterestCalculator.format_interest_rate(interest_rate)}")
                print(f"초기잔액: {BankUtils.format_currency(initial_balance)}")
                
                # 초기 입금 거래 기록
                if initial_balance > 0 and self.transaction_manager:
                    self.transaction_manager.record_deposit(
                        account_id, initial_balance, "계좌 개설", user_id
                    )
                
                return True
            else:
                print("❌ 계좌 생성에 실패했습니다.")
                return False
                
        except Exception as e:
            print(f"계좌 생성 오류: {e}")
            return False
    
    def save_account(self, account: Account) -> bool:
        """
        계좌 정보를 DB에 저장
        
        Args:
            account: 저장할 계좌 객체
            
        Returns:
            bool: 저장 성공 여부
        """
        try:
            result = self.db.execute_update(
                SQLQueries.INSERT_ACCOUNT,
                account.to_dict()
            )
            return result > 0
            
        except Exception as e:
            print(f"계좌 저장 오류: {e}")
            return False
    
    def list_accounts(self, user_id: str):
        """
        사용자의 계좌 목록 조회
        
        Args:
            user_id: 사용자 ID
        """
        try:
            print(f"\n[{self.user_manager.get_user_name(user_id)}님의 계좌 목록]")
            print("=" * 80)
            
            results = self.db.execute_query(
                SQLQueries.SELECT_ACCOUNTS_BY_USER,
                {'user_id': user_id}
            )
            
            if not results:
                print("등록된 계좌가 없습니다.")
                return
            
            print(f"{'계좌번호':<15} {'계좌명':<15} {'계좌종류':<10} {'잔액':<15} {'이자율':<10} {'개설일':<12}")
            print("-" * 80)
            
            for data in results:
                account = Account.from_dict(data)
                print(f"{account.account_id:<15} {account.account_name:<15} "
                      f"{account.account_type:<10} {BankUtils.format_currency(account.balance):<15} "
                      f"{InterestCalculator.format_interest_rate(account.interest_rate):<10} "
                      f"{account.create_date.strftime('%Y-%m-%d'):<12}")
            
            print("=" * 80)
            
        except Exception as e:
            print(f"계좌 목록 조회 오류: {e}")
    
    def read_account(self, user_id: str):
        """
        계좌 상세 조회
        
        Args:
            user_id: 사용자 ID
        """
        try:
            print("\n[계좌 상세 조회]")
            print("=" * 50)
            
            account_id = self.input_helper.input_account_id(
                "조회할 계좌번호: ", own_only=True, login_id=user_id
            )
            
            account = self.get_account_by_id(account_id)
            if not account:
                print("계좌를 찾을 수 없습니다.")
                return
            
            print(f"\n계좌번호: {account.account_id}")
            print(f"계좌명: {account.account_name}")
            print(f"계좌종류: {account.account_type}")
            print(f"잔액: {BankUtils.format_currency(account.balance)}")
            print(f"이자율: {InterestCalculator.format_interest_rate(account.interest_rate)}")
            print(f"개설일: {account.create_date.strftime('%Y-%m-%d %H:%M:%S')}")
            print(f"마지막 이자지급일: {account.last_interest_date.strftime('%Y-%m-%d %H:%M:%S')}")
            
        except Exception as e:
            print(f"계좌 조회 오류: {e}")
    
    def get_account_by_id(self, account_id: str) -> Optional[Account]:
        """
        계좌번호로 계좌 정보 조회
        
        Args:
            account_id: 계좌번호
            
        Returns:
            Optional[Account]: 계좌 객체 또는 None
        """
        try:
            results = self.db.execute_query(
                SQLQueries.SELECT_ACCOUNT_BY_ID,
                {'account_id': account_id}
            )
            
            if results:
                return Account.from_dict(results[0])
            
            return None
            
        except Exception as e:
            print(f"계좌 조회 오류: {e}")
            return None
    
    def delete_account_menu(self, user_id: str, account_id: Optional[str] = None):
        """
        계좌 해지 메뉴
        
        Args:
            user_id: 사용자 ID
            account_id: 해지할 계좌번호 (None이면 입력받음)
        """
        try:
            print("\n[계좌 해지]")
            print("=" * 50)
            
            if not account_id:
                account_id = self.input_helper.input_account_id(
                    "해지할 계좌번호: ", own_only=True, login_id=user_id
                )
            
            account = self.get_account_by_id(account_id)
            if not account:
                print("계좌를 찾을 수 없습니다.")
                return
            
            # 계좌 비밀번호 확인
            password = self.input_helper.input("계좌 비밀번호: ")
            if password != account.account_password:
                print("계좌 비밀번호가 일치하지 않습니다.")
                return
            
            # 잔액 확인
            if account.balance > 0:
                print(f"잔액이 {BankUtils.format_currency(account.balance)} 있습니다.")
                if not self.input_helper.confirm_action():
                    print("계좌 해지가 취소되었습니다.")
                    return
            
            # 계좌 해지 확인
            print(f"정말로 {account.account_name} 계좌를 해지하시겠습니까?")
            if not self.input_helper.confirm_action():
                print("계좌 해지가 취소되었습니다.")
                return
            
            # 계좌 해지 처리
            if self.delete_account(account_id):
                print("✅ 계좌가 해지되었습니다.")
            else:
                print("❌ 계좌 해지에 실패했습니다.")
                
        except Exception as e:
            print(f"계좌 해지 오류: {e}")
    
    def delete_account(self, account_id: str) -> bool:
        """
        계좌 삭제
        
        Args:
            account_id: 삭제할 계좌번호
            
        Returns:
            bool: 삭제 성공 여부
        """
        try:
            result = self.db.execute_update(
                SQLQueries.DELETE_ACCOUNT,
                {'account_id': account_id}
            )
            return result > 0
            
        except Exception as e:
            print(f"계좌 삭제 오류: {e}")
            return False
    
    def change_password(self, user_id: str):
        """
        계좌 비밀번호 변경
        
        Args:
            user_id: 사용자 ID
        """
        try:
            print("\n[계좌 비밀번호 변경]")
            print("=" * 50)
            
            account_id = self.input_helper.input_account_id(
                "비밀번호를 변경할 계좌번호: ", own_only=True, login_id=user_id
            )
            
            account = self.get_account_by_id(account_id)
            if not account:
                print("계좌를 찾을 수 없습니다.")
                return
            
            # 현재 비밀번호 확인
            current_password = self.input_helper.input("현재 계좌 비밀번호: ")
            if current_password != account.account_password:
                print("현재 비밀번호가 일치하지 않습니다.")
                return
            
            # 새 비밀번호 입력
            new_password = self.input_helper.input_account_password()
            
            # 비밀번호 변경 확인
            print("계좌 비밀번호를 변경하시겠습니까?")
            if not self.input_helper.confirm_action():
                print("비밀번호 변경이 취소되었습니다.")
                return
            
            # 비밀번호 업데이트
            if self.update_account_password(account_id, new_password):
                print("✅ 계좌 비밀번호가 변경되었습니다.")
            else:
                print("❌ 계좌 비밀번호 변경에 실패했습니다.")
                
        except Exception as e:
            print(f"계좌 비밀번호 변경 오류: {e}")
    
    def update_account_password(self, account_id: str, new_password: str) -> bool:
        """
        계좌 비밀번호 업데이트
        
        Args:
            account_id: 계좌번호
            new_password: 새 비밀번호
            
        Returns:
            bool: 업데이트 성공 여부
        """
        try:
            result = self.db.execute_update(
                SQLQueries.UPDATE_ACCOUNT_PASSWORD,
                {'account_id': account_id, 'account_password': new_password}
            )
            return result > 0
            
        except Exception as e:
            print(f"계좌 비밀번호 업데이트 오류: {e}")
            return False
    
    def update_account_balance(self, account_id: str, new_balance: float, 
                             last_interest_date: Optional[datetime] = None) -> bool:
        """
        계좌 잔액 업데이트
        
        Args:
            account_id: 계좌번호
            new_balance: 새 잔액
            last_interest_date: 마지막 이자 지급일
            
        Returns:
            bool: 업데이트 성공 여부
        """
        try:
            update_data = {
                'account_id': account_id,
                'balance': new_balance,
                'last_interest_date': last_interest_date or datetime.now()
            }
            
            result = self.db.execute_update(
                SQLQueries.UPDATE_ACCOUNT_BALANCE,
                update_data
            )
            return result > 0
            
        except Exception as e:
            print(f"계좌 잔액 업데이트 오류: {e}")
            return False

File: project/src/banksystem/manager/AccountManager.java
package banksystem.manager;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Scanner;
import java.util.Date;

import banksystem.entity.Account;
import banksystem.entity.Transaction;  // Transaction import 추가
import banksystem.entity.User;
import banksystem.helper.InputHelper;
import banksystem.helper.ValidationHelper;
import banksystem.util.BankUtils;
import banksystem.util.InterestCalculator;

public class AccountManager {
	private Connection conn;
	private ValidationHelper validator;
	private InputHelper inputHelper;
	private UserManager userManager;
	private Scanner scanner;
	private TransactionManager transactionManager; 

	public AccountManager(Connection conn, ValidationHelper validator, InputHelper inputHelper, UserManager userManager,
			Scanner scanner, TransactionManager transactionManager) {
		this.conn = conn;
		this.validator = validator;
		this.inputHelper = inputHelper;
		this.userManager = userManager;
		this.scanner = scanner;
		this.transactionManager = transactionManager; 
	}

	// Account 객체를 DB에 저장
	public boolean saveAccount(Account account) {
	    String sql = "INSERT INTO accounts (account_id, account_name, account_type, account_password, " +
	                 "balance, user_id, create_date, interest_rate, last_interest_date) " +
	                 "VALUES (?, ?, ?, ?, ?, ?, SYSDATE, ?, SYSDATE)";
	    try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
	        pstmt.setString(1, account.getAccountId());
	        pstmt.setString(2, account.getAccountName());
	        pstmt.setString(3, account.getAccountType());
	        pstmt.setString(4, account.getAccountPassword());
	        pstmt.setDouble(5, account.getBalance());
	        pstmt.setString(6, account.getUserId());
	        pstmt.setDouble(7, account.getInterestRate()); 
	        return pstmt.executeUpdate() > 0;
	    } catch (SQLException e) {
	        System.out.println("계좌 생성 오류: " + e.getMessage());
	        return false;
	    }
	}

	// Account ID로 Account 객체 조회
	public Account getAccountById(String accountId) {
		String sql = "SELECT * FROM accounts WHERE account_id = ?";
		try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
			pstmt.setString(1, accountId);
			try (ResultSet rs = pstmt.executeQuery()) {
				if (rs.next()) {
					Account account = new Account();
					account.setAccountId(rs.getString("account_id"));
					account.setAccountName(rs.getString("account_name"));
					account.setAccountType(rs.getString("account_type"));
					account.setAccountPassword(rs.getString("account_password"));
					account.setBalance(rs.getDouble("balance"));
					account.setUserId(rs.getString("user_id"));
					account.setCreateDate(rs.getDate("create_date"));
					return account;
				}
			}
		} catch (SQLException e) {
			System.out.println("계좌 조회 오류: " + e.getMessage());
		}
		return null;
	}

	// Account 잔액 업데이트
	public boolean updateAccountBalance(String accountId, double newBalance) {
		String sql = "UPDATE accounts SET balance = ? WHERE account_id = ?";
		try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
			pstmt.setDouble(1, newBalance);
			pstmt.setString(2, accountId);
			return pstmt.executeUpdate() > 0;
		} catch (SQLException e) {
			System.out.println("잔액 업데이트 오류: " + e.getMessage());
			return false;
		}
	}

	// Account 비밀번호 업데이트
	public boolean updateAccountPassword(String accountId, String newPassword) {
		String sql = "UPDATE accounts SET account_password = ? WHERE account_id = ?";
		try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
			pstmt.setString(1, newPassword);
			pstmt.setString(2, accountId);
			return pstmt.executeUpdate() > 0;
		} catch (SQLException e) {
			System.out.println("비밀번호 변경 오류: " + e.getMessage());
			return false;
		}
	}

	// Account 삭제
	public boolean deleteAccount(String accountId) {
		String sql = "DELETE FROM accounts WHERE account_id = ?";
		try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
			pstmt.setString(1, accountId);
			return pstmt.executeUpdate() > 0;
		} catch (SQLException e) {
			System.out.println("계좌 해지 오류: " + e.getMessage());
			return false;
		}
	}

	// 계좌 존재 여부 확인
	public boolean accountExists(String accountId) {
		String sql = "SELECT COUNT(*) FROM accounts WHERE account_id = ?";
		try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
			pstmt.setString(1, accountId);
			try (ResultSet rs = pstmt.executeQuery()) {
				return rs.next() && rs.getInt(1) > 0;
			}
		} catch (SQLException e) {
			System.out.println("계좌 조회 오류: " + e.getMessage());
		}
		return false;
	}

	// 본인 계좌 여부 확인
	public boolean isMyAccount(String accountId, String loginId) {
		Account account = getAccountById(accountId);
		return account != null && loginId != null && loginId.equals(account.getUserId());
	}

	// 계좌 비밀번호 검증
	public boolean verifyPassword(String accountId, String password) {
		Account account = getAccountById(accountId);
		return account != null && password.equals(account.getAccountPassword());
	}

	// 계좌 소유자명 조회
	public String getAccountHolderName(String accountId) {
		String sql = "SELECT u.user_name FROM accounts a " + "JOIN users u ON a.user_id = u.user_id "
				+ "WHERE a.account_id = ?";
		try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
			pstmt.setString(1, accountId);
			try (ResultSet rs = pstmt.executeQuery()) {
				if (rs.next()) {
					return rs.getString("user_name");
				}
			}
		} catch (SQLException e) {
			System.out.println("예금주명 조회 오류: " + e.getMessage());
		}
		return "미상";
	}

	// 계좌 잔액 조회
	public double getBalance(String accountId) {
		Account account = getAccountById(accountId);
		return account != null ? account.getBalance() : -1;
	}

	// 계좌 생성
	public void createAccount(String loginId) {
		System.out.println("[계좌 신규 개설]");
		String[] types = { "보통예금", "정기예금", "적금" };

		System.out.println("\n[계좌 종류 선택]");
		System.out.println("============================================================================================================================");
		System.out.println("1. 보통예금 - [연 0.1%]");
		System.out.println("2. 정기예금 - [연 1.5%]");
		System.out.println("3. 적   금 - [연 2.0%]");
		System.out.println("============================================================================================================================");
		int choice;
		do {
			try {
				System.out.print("계좌 종류 선택 (1 - 3): ");
				choice = Integer.parseInt(scanner.nextLine());
				if (choice >= 1 && choice <= 3)
					break;
			} catch (NumberFormatException e) {
			}
			System.out.println("1 ~ 3번을 선택해주세요.");
		} while (true);

		String accountType = types[choice - 1];
		String accountName = accountType + " 계좌_" + userManager.getUserName(loginId);
		String password = inputHelper.inputAccountPassword();
		double initialBalance = inputHelper.inputAmount("초기 입금액 (1,000원 이상): ");

		if (inputHelper.confirmAction()) {
			String accountId = BankUtils.generateAccountNumber(conn);

			// 계좌 종류에 따른 이자율 자동 설정
			double interestRate = InterestCalculator.getInterestRateByType(accountType);
			
			// Account 객체 생성
			Account account = new Account(accountId, accountName, accountType, password, initialBalance, loginId);
			account.setInterestRate(interestRate); // 이자율 설정
	        account.setLastInterestDate(new Date()); // 마지막 이자 지급일을 현재 날짜로 설정
			
			// 트랜잭션으로 계좌 생성과 초기 입금 거래내역을 함께 처리
			try {
				conn.setAutoCommit(false);  // 트랜잭션 시작
				
				// 1. 계좌 생성
				if (saveAccount(account)) {
					// 2. 초기 입금 거래내역 생성 및 저장
					if (transactionManager != null) {
						Transaction initialTransaction = new Transaction();
						initialTransaction.setTransactionId(BankUtils.generateTransactionId(conn));
						initialTransaction.setAccountId(accountId);
						initialTransaction.setTransactionType("입금");
						initialTransaction.setAmount(initialBalance);
						initialTransaction.setBalanceAfter(initialBalance);
						initialTransaction.setTransactionMemo("계좌 개설 초기 입금");
						
						// TransactionManager를 통해 거래내역 저장
						if (transactionManager.saveTransaction(initialTransaction)) {
							conn.commit();  // 모두 성공시 커밋
							
							System.out.println("✅ 계좌가 성공적으로 개설되었습니다!");
							System.out.println("   계좌번호: " + accountId);
							System.out.println("   계좌종류: " + accountType);
							System.out.println("   적용 이자율: " + String.format("%.1f%%", interestRate * 100));
						} else {
							throw new SQLException("거래내역 저장 실패");
						}
					} else {
						// TransactionManager가 null이면 계좌만 생성
						conn.commit();
						System.out.println("✅ 계좌가 성공적으로 개설되었습니다!");
						System.out.println("   계좌번호: " + accountId);
						System.out.println("   계좌종류: " + accountType);
						System.out.println("   적용 이자율: " + String.format("%.1f%%", interestRate * 100));
					}
				} else {
					throw new SQLException("계좌 생성 실패");
				}
				
			} catch (SQLException e) {
				try {
					conn.rollback();  // 실패시 롤백
					System.out.println("계좌 개설 중 오류가 발생했습니다: " + e.getMessage());
				} catch (SQLException ex) {
					System.out.println("롤백 오류: " + ex.getMessage());
				}
			} finally {
				try {
					conn.setAutoCommit(true);  // 자동커밋 복원
				} catch (SQLException e) {
					System.out.println("자동커밋 설정 오류: " + e.getMessage());
				}
			}
		}
	}

	// 계좌 조회
	public void readAccount(String loginId) {
		System.out.println("[계좌 조회]");
		String accountId = inputHelper.input("계좌번호: ");

		Account account = getAccountById(accountId);
		if (account != null) {
			User accountHolder = userManager.getUserById(account.getUserId());

			System.out.println("계좌번호: " + account.getAccountId());
			System.out.println("계좌명: " + account.getAccountName());
			System.out.println("계좌종류: " + account.getAccountType());
			System.out.println("잔액: " + BankUtils.formatCurrency(account.getBalance()));
			System.out.println("소유자: " + (accountHolder != null ? accountHolder.getUserName() : "미상"));
			System.out.println("개설일: " + account.getCreateDate());
	        
	        // 메뉴로 돌아가기
	        System.out.println("\n엔터키를 누르면 메뉴로 돌아갑니다.");
	        scanner.nextLine();
	        
		} else {
			System.out.println("해당 계좌를 찾을 수 없습니다.");
	        System.out.println("\n엔터키를 누르면 메뉴로 돌아갑니다.");
	        scanner.nextLine();
		}
	}

	// 계좌 삭제
	public void deleteAccountMenu(String loginId, String accountId) {
		if (accountId == null) {
			System.out.println("[계좌 해지]");
			accountId = inputHelper.inputAccountId("해지할 계좌번호: ", true, loginId);
		}

		if (!checkPassword(accountId) || !inputHelper.confirmAction()) {
			return;
		}

		if (deleteAccount(accountId)) {
			System.out.println("✅ 계좌가 해지되었습니다!");
		}
	}

	// 계좌 비밀번호 변경
	public void changePassword(String loginId) {
		System.out.println("[계좌 비밀번호 변경]");
		String accountId = inputHelper.inputAccountId("계좌번호: ", true, loginId);

		if (checkPassword(accountId)) {
			// 현재 계좌 정보 조회
			Account account = getAccountById(accountId);
			String currentPassword = account != null ? account.getAccountPassword() : null;
			String newPassword;

			do {
				System.out.print("새 계좌 비밀번호 (4자리 숫자): ");
				newPassword = scanner.nextLine();

				// 유효성 검사
				if (!validator.validateAccountPassword(newPassword)) {
					continue; // 유효성 검사 실패시 다시 입력
				}

				if (currentPassword != null && currentPassword.equals(newPassword)) {
					System.out.println("새 비밀번호는 현재 비밀번호와 달라야 합니다.");
				} else {
					break;
				}
			} while (true);

			if (inputHelper.confirmAction()) {
				if (updateAccountPassword(accountId, newPassword)) {
					System.out.println("✅ 계좌 비밀번호가 변경되었습니다!");
				}
			}
		}
	}

	public void listAccounts(String loginId) {
		System.out.println("\n[계좌 목록] 사용자: " + userManager.getUserName(loginId) + " (" + loginId + ")");
		
		if (transactionManager != null) {
			System.out.println("============================================================================================================================");
			System.out.println();
			String BOLD = "\033[1m";
			String RESET = "\033[0m";
			System.out.println("💡 " + BOLD + "거래 한도 안내" + RESET);
			
			String[] limits = transactionManager.getTransactionLimits();
			for (String limit : limits) {
				System.out.println(limit);
			}
		}
		System.out.println();
		System.out.println("============================================================================================================================");
		System.out.println("계좌번호\t\t계좌명\t\t\t계좌종류\t\t소유자\t\t잔액");
		System.out.println("============================================================================================================================");

		String sql = "SELECT a.*, u.user_name FROM accounts a JOIN users u ON a.user_id = u.user_id "
				+ "WHERE a.user_id = ?";

		double totalBalance = 0; // 전체 잔액 합계 변수

		try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
			pstmt.setString(1, loginId);
			try (ResultSet rs = pstmt.executeQuery()) {
				boolean hasAccounts = false;
				while (rs.next()) {
					hasAccounts = true;

					String accountId = rs.getString("account_id");
					String accountName = rs.getString("account_name");
					String displayAccountName = accountName.length() > 12 ? accountName.substring(0, 12) + ".."
							: accountName;

					double balance = rs.getDouble("balance");
					totalBalance += balance; // 각 계좌 잔액을 합계에 누적

					System.out.println(accountId + "\t" + displayAccountName + "\t\t"
							+ rs.getString("account_type") + "\t\t" + rs.getString("user_name") + "\t\t"
							+ BankUtils.formatCurrency(balance));
					
					if (transactionManager != null) {
						String usageInfo = transactionManager.getFormattedUsageByAccount(accountId);
						System.out.println(usageInfo);
						System.out.println();
					}
				}

				if (!hasAccounts) {
					System.out.println("보유하신 계좌가 없습니다. 계좌를 개설하세요!");
				} else {
					// 전체 잔액 합계 표시 (오른쪽 정렬)
					System.out.println("============================================================================================================================");
					System.out.println("\t\t\t\t\t\t\t"+"전체 계좌 잔액 합계: " + BankUtils.formatCurrency(totalBalance));
				}
			}
		} catch (SQLException e) {
			System.out.println("계좌 목록 조회 오류: " + e.getMessage());
		}
	}

	// 계좌 비밀번호 확인 (InputHelper에서 사용)
	public boolean checkPassword(String accountId) {
        String password;
        do {
            System.out.print("계좌 비밀번호 (4자리 숫자): ");
            password = scanner.nextLine();
            if (password.length() != 4 || !password.matches("\\d{4}")) {
                System.out.println("계좌 비밀번호는 4자리 숫자여야 합니다.");
                continue;
            }
            if (verifyPassword(accountId, password))
                return true;
            System.out.println("계좌 비밀번호가 일치하지 않습니다.");
        } while (true);
    }
}